File: android/app/src/main/kotlin/com/mineracaomg/GeoJSONService.kt
package com.mineracaomg

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import org.osmdroid.util.GeoPoint
import java.net.URL

class GeoJSONService {

    // Fetches polygon rings for all 12 mesorregiões in parallel.
    // Returns: regionId → list of rings (each ring is a list of GeoPoint).
    suspend fun fetchAllRegions(): Map<String, List<List<GeoPoint>>> = coroutineScope {
        val deferreds = Region.all.map { region ->
            async { region.id to fetchRegion(region.ibgeCode) }
        }
        deferreds.associate { it.await() }
    }

    // Fetches all rings of the MG state boundary (used for the inverse mask).
    suspend fun fetchStateBoundary(): List<List<GeoPoint>> = withContext(Dispatchers.IO) {
        val url = "https://servicodados.ibge.gov.br/api/v3/malhas/estados/31" +
                "?formato=application%2Fvnd.geo%2Bjson"
        val json = URL(url).openStream().bufferedReader().readText()
        parseAllRings(json)
    }

    private suspend fun fetchRegion(ibgeCode: Int): List<List<GeoPoint>> = withContext(Dispatchers.IO) {
        val url = "https://servicodados.ibge.gov.br/api/v3/malhas/mesorregioes/$ibgeCode" +
                "?formato=application%2Fvnd.geo%2Bjson"
        val json = URL(url).openStream().bufferedReader().readText()
        parseAllRings(json)
    }

    // Parses all outer rings from a GeoJSON FeatureCollection (Polygon or MultiPolygon).
    private fun parseAllRings(geojson: String): List<List<GeoPoint>> {
        val rings = mutableListOf<List<GeoPoint>>()
        try {
            val root = JSONObject(geojson)
            val features = root.getJSONArray("features")
            for (i in 0 until features.length()) {
                val geometry = features.getJSONObject(i).getJSONObject("geometry")
                val type = geometry.getString("type")
                val coords = geometry.getJSONArray("coordinates")
                when (type) {
                    "Polygon" -> {
                        rings.add(parseRing(coords.getJSONArray(0)))
                    }
                    "MultiPolygon" -> {
                        for (pi in 0 until coords.length()) {
                            rings.add(parseRing(coords.getJSONArray(pi).getJSONArray(0)))
                        }
                    }
                }
            }
        } catch (e: Exception) {
            android.util.Log.e("GeoJSONService", "Parse error", e)
        }
        return rings
    }

    private fun parseRing(coordArray: JSONArray): List<GeoPoint> {
        val points = mutableListOf<GeoPoint>()
        for (i in 0 until coordArray.length()) {
            val pt = coordArray.getJSONArray(i)
            val lon = pt.getDouble(0)
            val lat = pt.getDouble(1)
            points.add(GeoPoint(lat, lon))
        }
        return points
    }
}

File: ios/MineracaoMG/Services/GeoJSONService.swift
import Foundation
import MapKit

// MARK: - GeoJSONService
// Fetches mesorregião GeoJSON from IBGE API v3 and converts to MKPolygon arrays.

final class GeoJSONService {

    static let shared = GeoJSONService()
    private init() {}

    // One polygon array per region ID
    typealias RegionPolygons = [String: [MKPolygon]]

    /// Fetches all 12 mesorregião polygons concurrently.
    func fetchAllRegions() async throws -> RegionPolygons {
        var result: RegionPolygons = [:]
        try await withThrowingTaskGroup(of: (String, [MKPolygon]).self) { group in
            for region in Region.all {
                group.addTask {
                    let polygons = try await self.fetchRegion(region)
                    return (region.id, polygons)
                }
            }
            for try await (id, polygons) in group {
                result[id] = polygons
            }
        }
        return result
    }

    /// Fetches the MG state boundary for building the mask polygon.
    func fetchStateBoundary() async throws -> [MKPolygon] {
        let url = URL(string: "https://servicodados.ibge.gov.br/api/v3/malhas/estados/31?formato=application/vnd.geo%2Bjson")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return polygons(from: data, regionId: "boundary")
    }

    // MARK: Private

    private func fetchRegion(_ region: Region) async throws -> [MKPolygon] {
        let url = URL(string: "https://servicodados.ibge.gov.br/api/v3/malhas/mesorregioes/\(region.ibgeCode)?formato=application/vnd.geo%2Bjson")!
        let (data, _) = try await URLSession.shared.data(from: url)
        return polygons(from: data, regionId: region.id)
    }

    private func polygons(from data: Data, regionId: String) -> [MKPolygon] {
        guard let fc = try? JSONDecoder().decode(GeoJSONFeatureCollection.self, from: data) else {
            return []
        }
        var result: [MKPolygon] = []
        for feature in fc.features {
            switch feature.geometry {
            case .polygon(let rings):
                if let poly = mkPolygon(rings: rings, title: regionId) {
                    result.append(poly)
                }
            case .multiPolygon(let polys):
                for rings in polys {
                    if let poly = mkPolygon(rings: rings, title: regionId) {
                        result.append(poly)
                    }
                }
            }
        }
        return result
    }

    private func mkPolygon(rings: [[[Double]]], title: String) -> MKPolygon? {
        guard let exterior = rings.first, !exterior.isEmpty else { return nil }
        var outerCoords = exterior.compactMap { c -> CLLocationCoordinate2D? in
            guard c.count >= 2 else { return nil }
            return CLLocationCoordinate2D(latitude: c[1], longitude: c[0])
        }
        let interiors: [MKPolygon] = rings.dropFirst().compactMap { ring in
            var coords = ring.compactMap { c -> CLLocationCoordinate2D? in
                guard c.count >= 2 else { return nil }
                return CLLocationCoordinate2D(latitude: c[1], longitude: c[0])
            }
            guard !coords.isEmpty else { return nil }
            return MKPolygon(coordinates: &coords, count: coords.count)
        }
        let polygon = MKPolygon(coordinates: &outerCoords, count: outerCoords.count,
                                interiorPolygons: interiors.isEmpty ? nil : interiors)
        polygon.title = title
        return polygon
    }
}

// MARK: - GeoJSON Decodable models

private struct GeoJSONFeatureCollection: Decodable {
    let features: [GeoJSONFeature]
}

private struct GeoJSONFeature: Decodable {
    let geometry: GeoJSONGeometry
}

private enum GeoJSONGeometry: Decodable {
    case polygon([[[Double]]])
    case multiPolygon([[[[Double]]]])

    private enum CodingKeys: String, CodingKey { case type, coordinates }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        let type = try c.decode(String.self, forKey: .type)
        switch type {
        case "Polygon":
            self = .polygon(try c.decode([[[Double]]].self, forKey: .coordinates))
        case "MultiPolygon":
            self = .multiPolygon(try c.decode([[[[Double]]]].self, forKey: .coordinates))
        default:
            throw DecodingError.dataCorruptedError(forKey: .type, in: c, debugDescription: "Unknown geometry type: \(type)")
        }
    }
}
